usr/src/uts/common/fs/zfs/vdev.c

```diff
@@ -3463,3 +3480,114 @@ vdev_deadman(vdev_t *vd)
 		mutex_exit(&vq->vq_lock);
 	}
 }
+
+/*
+ * Implements the per-vdev portion of manual TRIM. The function passes over
+ * all metaslabs on this vdev and performs a metaslab_trim_all on them. It's
+ * also responsible for rate-control if spa_man_trim_rate is non-zero.
+ */
+void
+vdev_man_trim(vdev_trim_info_t *vti)
+{
+	clock_t t = ddi_get_lbolt();
+	spa_t *spa = vti->vti_vdev->vdev_spa;
+	vdev_t *vd = vti->vti_vdev;
+	uint64_t cursor;
+	boolean_t was_loaded = B_FALSE;
+
+	vd->vdev_man_trimming = B_TRUE;
+	vd->vdev_trim_prog = 0;
+
+	spa_config_enter(spa, SCL_STATE_ALL, FTAG, RW_READER);
+	ASSERT(vd->vdev_ms[0] != NULL);
+	cursor = vd->vdev_ms[0]->ms_start;
+	for (uint64_t i = 0; i < vti->vti_vdev->vdev_ms_count &&
+	    !spa->spa_man_trim_stop;) {
+		uint64_t delta;
+		metaslab_t *msp = vd->vdev_ms[i];
+		zio_t *trim_io;
+
+		trim_io = metaslab_trim_all(msp, &cursor, &delta, &was_loaded);
+		spa_config_exit(spa, SCL_STATE_ALL, FTAG);
+
+		if (trim_io != NULL) {
+			ASSERT3U(cursor, >=, vd->vdev_ms[0]->ms_start);
+			vd->vdev_trim_prog = cursor - vd->vdev_ms[0]->ms_start;
+			(void) zio_wait(trim_io);
+		} else {
+			/*
+			 * If there was nothing more left to trim, that means
+			 * this metaslab is either done trimming, or we
+			 * couldn't load it, move to the next one.
+			 */
+			i++;
+			if (i < vti->vti_vdev->vdev_ms_count)
+				ASSERT3U(vd->vdev_ms[i]->ms_start, ==, cursor);
+		}
+
+		/* delay loop to handle fixed-rate trimming */
+		for (;;) {
+			uint64_t rate = spa->spa_man_trim_rate;
+			uint64_t sleep_delay;
+
+			if (rate == 0) {
+				/* No delay, just update 't' and move on. */
+				t = ddi_get_lbolt();
+				break;
+			}
+
+			sleep_delay = (delta * hz) / rate;
+			mutex_enter(&spa->spa_man_trim_lock);
+			(void) cv_timedwait(&spa->spa_man_trim_update_cv,
+			    &spa->spa_man_trim_lock, t);
```

Comment:
This will never delay.  Here's my proposed patch:
```
diff --git a/module/zfs/vdev.c b/module/zfs/vdev.c
index ca85120..ca2ba5c 100644
--- a/module/zfs/vdev.c
+++ b/module/zfs/vdev.c
@@ -3772,6 +3772,7 @@ vdev_man_trim(vdev_trim_info_t *vti)
                for (;;) {
                        uint64_t rate = spa->spa_man_trim_rate;
                        uint64_t sleep_delay;
+                       clock_t t1;
 
                        if (rate == 0) {
                                /* No delay, just update 't' and move on. */
@@ -3781,8 +3782,8 @@ vdev_man_trim(vdev_trim_info_t *vti)
 
                        sleep_delay = (delta * hz) / rate;
                        mutex_enter(&spa->spa_man_trim_lock);
-                       (void) cv_timedwait(&spa->spa_man_trim_update_cv,
-                           &spa->spa_man_trim_lock, t);
+                       t1 = cv_timedwait(&spa->spa_man_trim_update_cv,
+                           &spa->spa_man_trim_lock, t + sleep_delay);
                        mutex_exit(&spa->spa_man_trim_lock);
 
                        /* If interrupted, don't try to relock, get out */
@@ -3790,7 +3791,7 @@ vdev_man_trim(vdev_trim_info_t *vti)
                                goto out;
 
                        /* Timeout passed, move on to the next metaslab. */
-                       if (ddi_get_lbolt() >= t + sleep_delay) {
+                       if (t1 == -1) {
                                t += sleep_delay;
                                break;
                        }
```